File: coding/usaco/milk2/milk2.c
/*
ID: rohitya1
LANG: C
PROG: milk2
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// took 40 mins to think, need to improve timings
int main() {
    FILE* fin = fopen("milk2.in", "r");
    FILE* fout = fopen("milk2.out", "w");

    int n;
    int time[5000][2];
    int minTime = 0;
    int maxTime = 0;
    int milkTime = 0;
    int nonMilkTime = 0;
    int i, j, start, stop;

    fscanf(fin, "%d\n", &n);
    for (i = 0; i < n; i++) //C99 loop
        fscanf(fin, "%d %d\n", &time[i][0], &time[i][1]);

    //sort time by time[0]
    for (i = 0; i < n-1; i++) {
        for (j = i+1; j < n; j++) {
            if (time[j][0] < time[i][0]) {
                start = time[i][0];
                stop = time[i][1];
                // swap i and j
                time[i][0] = time[j][0];
                time[i][1] = time[j][1];

                time[j][0] = start;
                time[j][1] = stop;
            }
        }
    }

    minTime = time[0][0];
    maxTime = time[0][1];
    milkTime = maxTime - minTime;
    nonMilkTime = 0;
    for (i = 1; i < n; i++) {
        // no overlap
        if (time[i][0] > maxTime) {
            minTime = time[i][0];
            if ((time[i][0]-maxTime) > nonMilkTime)
                nonMilkTime = (time[i][0] - maxTime);
            maxTime = time[i][1];
        } else {
            if (time[i][1] > maxTime) {
                maxTime = time[i][1];
            }
        }
        if ((maxTime - minTime) > milkTime)
            milkTime = maxTime - minTime;
    }

    fprintf(fout, "%d %d\n", milkTime, nonMilkTime);

    fclose(fin);
    fclose(fout);
    return 0;
}

File: coding/usaco/milk2/milk2.java
/*
ID: rohitya1
LANG: JAVA
PROG: milk2
*/

import java.util.*;
import java.lang.*;
import java.io.*;

class event {
    public int time; // time after 5H
    public boolean status; //start = true, stop = falsea
    public event(int time, boolean status) {
        this.time = time;
        this.status = status;
    }
}

class milk2 {
    public static void main(String[] args) throws IOException {
        BufferedReader in = new BufferedReader(new FileReader("milk2.in"));
        PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter("milk2.out")));

        int n = Integer.parseInt(in.readLine());
        List<event> events = new ArrayList<event>();

        for (int i = 0; i < n; i++) {
            StringTokenizer ss = new StringTokenizer(in.readLine());
            events.add(new event(Integer.parseInt(ss.nextToken()), true));
            events.add(new event(Integer.parseInt(ss.nextToken()), false));
        }

        Comparator<event> cmp = new Comparator<event>(){
            public int compare(event a, event b) {
                return Integer.valueOf(a.time).compareTo(Integer.valueOf(b.time));
            }
        };

        Collections.sort(events, cmp);

        int milkers = 0;
        int milkTime = 0;
        int nonMilkTime = 0;
        int start = events.get(0).time;
        int last = start;
        boolean noneBefore = false;

        for (event e: events) {
            if (e.status) milkers++;
            else milkers--;

            if (milkers == 1 && noneBefore && (e.time != last)) {
                start = e.time;
                int length = e.time - last;
                if (length > nonMilkTime) nonMilkTime = length;
            }

            if (milkers == 0) {
                int length = e.time - start;
                if (length > milkTime) milkTime = length;
            }

            if (milkers == 0) noneBefore = true;
            else noneBefore = false;

            last = e.time;
        }

        out.printf("%d %d\n", milkTime, nonMilkTime);

        in.close();
        out.close();

        System.exit(0);
    }
}
